Fixes #13222 - Shows pulp status and storage in proxy#show

shlomizadok opened the pull request:
Works with https://github.com/theforeman/foreman/pull/3069 and smart-proxy-pulp https://github.com/theforeman/smart-proxy-pulp/commit/7b4d074b904bedaa595ac53986f579bff14882b8

Todo 
- [x] Add tests

Would love to hear your comments meanwhile

xprazak2 commented on app/services/proxy_status/pulp.rb:
```diff
@@ -0,0 +1,46 @@
+    end
+  end
+end
+# ::ProxyStatus.status_registry.add(ProxyStatus::Pulp)
```
I get:
`
NoMethodError (undefined method 'status' for nil:NilClass): /home/oprazak/projects/katello/app/controllers/katello/concerns/smart_proxies_controller_extensions.rb:20:in 'pulp_status'
`
this is probably the reason. Uncommenting does not work for me, so maybe move it to Engine?

xprazak2 replied:
`self.humanized_name` is not defined in `ProxyStatus::Pulp`

xprazak2 commented:
When pulp storage cannot be detected, the spinner in Overview tab spins indefinitely.
![image](http://i.imgur.com/JXUW10W.png)

ohadlevy commented on app/services/proxy_status/pulp.rb:
```diff
@@ -0,0 +1,46 @@
+
+    def storage
+      fetch_proxy_data do
+        api.pulp_storage
```
is this not cached by purpose ?

ohadlevy commented on app/views/smart_proxies/_disk_usage.html.erb:
```diff
@@ -0,0 +1,9 @@
+</div>
+<div class="progress progress-label-left">
+  <div class="progress-bar" role="progressbar" aria-valuenow="<%= disk['percent'].delete('%') %>" aria-valuemin="0" aria-valuemax="100" style="width: <%= disk['percent'] %>;">
+    <span><%= disk['percent'] %></span>
```
should we change color if its above a certain percentage ?

xprazak2 commented:
The total space does not really show for me:
![img](http://i.imgur.com/GnjIC6e.png)

ohadlevy commented:
yeah, it should be disk['total']

ehelms commented on app/controllers/katello/concerns/smart_proxies_controller_extensions.rb:
```diff
@@ -0,0 +1,45 @@
+        before_filter :find_resource_and_status, :only => [:pulp_storage, :pulp_status]
+        alias_method_chain :action_permission, :katello
+
+        def pulp_storage
```
I am not as familiar with the design of this area (feel free to point me at them), but if I read this correctly, the intent appears to be to add these actions for showing showing status in the smart proxy UI but not to also enable this for retrieval in the API?

shlomizadok replied:
https://dl.dropboxusercontent.com/u/5892944/Capsule-2015-11-02.pdf (page 3)
I'd guess API representation would be appreciated

ehelms replied:
To me it would -- as I don't know why we would want to introduce some data as a UI only thing. Granted, I am also used to the Katello UI which inevitably requires APIs be built to support it (two birds one stone). If APIs are not in scope for this just let me know - some stories to track creating them would be nice.

ohadlevy replied:
@ehels API ATM is secondary, (phase two) one of the main reason its split
into its own service class is to reuse via the api controllers later on.

ehelms commented on app/helpers/katello/concerns/smart_proxy_helper_extensions.rb:
```diff
@@ -0,0 +1,28 @@
+      def boolean_to_icon(boolean)
+        boolean = Foreman::Cast.to_bool(boolean)
+        boolean ?  icon_text('ok', '', {:kind => 'pficon'}) :
+        icon_text('error-circle-o', '', {:kind => 'pficon'})
```
Splitting this across two lines without indentation is confusing to read. You could also consider shortening it to something like:

```
icon  = boolean ? 'ok' : 'error-circle-o'
icon_text(icon, '', {:kind => 'pficon'})
```

ehelms commented on app/lib/proxy_api/pulp.rb:
```diff
@@ -0,0 +1,19 @@
+module ProxyAPI
```
In the Foreman code base, these lives under 'lib/proxy_api' -- could we move it to there in this code base for consistency?

ehelms commented on app/services/proxy_status/pulp.rb:
```diff
@@ -0,0 +1,46 @@
+      end
+
+      json
+    end
```
This introduces some code overlap with https://github.com/Katello/katello/blob/master/app/models/katello/ping.rb#L105 -- could you consider re-use ?

shlomizadok replied:
I was actually thinking of doing the other way around (e.g., to create a separate PR to refactor https://github.com/Katello/katello/blob/master/app/models/katello/ping.rb#L105 to use this service class) :smile: 
Let me know which way you'd prefer.

ehelms replied:
I'm good with that approach.

On Tue, Jan 19, 2016 at 7:09 AM, Shlomi Zadok <notifications@github.com>
wrote:

> In app/services/proxy_status/pulp.rb
> <https://github.com/Katello/katello/pull/5716#discussion_r50104629>:
>
> > +      json['errors'] = {}
> > +
> > +      if json['known_workers'].empty?
> > +        json['errors']['known_workers'] = _("No pulp workers running.")
> > +      end
> > +
> > +      if json['database_connection'] && json['database_connection']['connected'] != true
> > +        json['errors']['database_connection'] = _("Pulp database connection issue.")
> > +      end
> > +
> > +      if json['messaging_connection'] && json['messaging_connection']['connected'] != true
> > +        json['errors']['messaging_connection'] = _("Pulp message bus connection issue.")
> > +      end
> > +
> > +      json
> > +    end
>
> I was actually thinking of doing the other way around (e.g., to create a
> separate PR to refactor
> https://github.com/Katello/katello/blob/master/app/models/katello/ping.rb#L105
> to use this service class) [image: :smile:]
> Let me know which way you'd prefer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Katello/katello/pull/5716/files#r50104629>.
>



-- 
Eric D. Helms
Red Hat Engineering
Ph.D. Student - North Carolina State University

ehelms commented on app/views/smart_proxies/pulp_status.html.erb:
```diff
@@ -0,0 +1,26 @@
+<%# @pulp_status %>
```
Is this needed here?

ehelms commented on app/helpers/katello/concerns/smart_proxy_helper_extensions.rb:
```diff
@@ -0,0 +1,28 @@
+          values['total'] = values.delete('1g-blocks')
+          disks << values
+
+
```
These are unneeded, however, using some blank lines to split up the logical code chunks above would be nice for readability.

jlsherrill commented on lib/katello/plugin.rb:
```diff
@@ -164,4 +164,11 @@
+    cx.add_pagelet :details_content,
+                   :name => _('Storage'),
+                   :partial => 'smart_proxies/show/storage',
+                   :onlyif => Proc.new { |proxy| proxy.has_feature?(SmartProxy::PULP_FEATURE) }
```
You probably want to check for the PULP_NODE_FEATURE feature as well.  We use that to differentiate between the 'built in' capsule and remote capsules.

shlomizadok commented:
Still a [WIP] but now can be tested. @ehelms, @jlsherrill - would love to have your comments

xprazak2 commented on app/views/smart_proxies/_disk_usage.html.erb:
```diff
@@ -0,0 +1,9 @@
+<%= disk['header'] %>
+<div class="storage-data">
+  <span class="big"><%= disk['used'] %></span> GB of <%= disk['total'] %> GB
+</div>
```
Missing span?
`<span class="big"><%= disk['total'] %></span> `

shlomizadok replied:
Don't think so, frameworks shows
![image](https://cloud.githubusercontent.com/assets/433583/12419028/1b13ced2-bebc-11e5-8b37-52cdc0e50525.png)

shlomizadok replied:
It should be i18n though :)

xprazak2 replied:
Ahh, I see now. I apologize, please disregard my previous comment.

xprazak2 commented on app/controllers/katello/concerns/smart_proxies_controller_extensions.rb:
```diff
@@ -0,0 +1,45 @@
+        end
+
+        def pulp_status
+          @pulp_status = @proxy_status[:pulp].status
```
When storage cannot be detected, spinner spins indefinitely. Maybe we could handle the exception here and render appropriate json.

shlomizadok commented:
Current UI:
![image](https://cloud.githubusercontent.com/assets/433583/12430662/7b7fab82-bef9-11e5-97f2-994f42965db8.png)

And 
![image](https://cloud.githubusercontent.com/assets/433583/12430672/8ac7a0ae-bef9-11e5-8e72-0311f919438e.png)

ehelms commented:
Seeing the following and a spinner on Pulp Storage:

```
2016-01-16T17:06:47 [app] [F]
 | Foreman::WrappedException (ERF50-5345 [Foreman::WrappedException]: Unable to connect ([ProxyAPI::ProxyException]: ERF12-3580 [ProxyAPI::ProxyException]: Unable to detect pulp storage ([RestClient::Reso...)):
 |   app/services/proxy_status/base.rb:40:in `rescue in fetch_proxy_data'
```

Not sure if I am missing something setup wise.

jlsherrill commented:
I'm seeing the same error as eric, at first i hadn't updated my local copy of the smart-proxy-pulp plugin, but even after that i was seeing the same error. seems that the smart proxy is throwing a 404:

#<ProxyAPI::ProxyException: ERF12-3580 [ProxyAPI::ProxyException]: Unable to detect pulp storage ([RestClient::ResourceNotFound]: 404 Resource Not Found) for proxy https://katello-centos7-devel.example.com:9090/pulp/status/disk_usage?size=gigabyte>

I couldn't quite figure out why, and actually after the installation it doesn't seem to be loading my pulp plugin properly at all, a curl of https://katello-centos7-devel.example.com:9090/features no longer shows pulp.  Will investigate

jlsherrill commented:
It actually seems like once i apply https://github.com/theforeman/smart-proxy-pulp/commit/7b4d074b904bedaa595ac53986f579bff14882b8, my install of smart-proxy will refuse to load the pulp plugin without any error

jlsherrill commented:
ahha!   after hacking the smart proxy code i see:

smart-proxy: Couldn't enable plugin pulp: File at '/var/lib/pulp/content' defined in 'pulp_content_dir' parameter doesn't exist or is unreadableErrors detected on startup,

Not sure it wasn't being printed even though there were log statements to catch this.

As it turns out /var/lib/pulp/content does not yet exist because i haven't synced any content yet. Yay!  

I'm guessing  @ehelms spun up a new dev environment and was also hitting this.  You probably need to handle this case where a particular directory doesn't exist (or isn't readable) in the smart proxy pulp plugin.

After syncing some content (to make sure that the directory is created properly), i see both the status and the storage.

jlsherrill commented on app/helpers/katello/concerns/smart_proxy_helper_extensions.rb:
```diff
@@ -0,0 +1,28 @@
+          mount_points[mount] << name
+          values['header'] = "#{mount_points[mount].to_sentence} (#{values['path']} on #{values['filesystem']})"
+          values['total'] = values.delete('1g-blocks')
+          disks << values
```
just wanted to :+1: this logic.  Initially i thought you were adding all the diskspace together, but consolidating among file systems is very clever and very useful to the user.

jlsherrill replied:
I do see in the UI:

pulp_dir, pulp_content_dir, and mongodb_dir (/var/lib/mongodb on /dev/vda3)

but i might expect the list of directories rather than their config values.

shlomizadok commented:
With latest commit UI looks like:
![image](https://cloud.githubusercontent.com/assets/433583/12448867/9a151e82-bf81-11e5-86f4-5293e533a511.png)

and:
![image](https://cloud.githubusercontent.com/assets/433583/12448939/4350ac32-bf82-11e5-9522-1578b38684cd.png)

shlomizadok commented:
As for pulp on smart proxy, we had a discussion on the pulp-smart-proxy PR and we thought it would be wise not to start the pulp plugin on the proxy if those three directories (/var/lib/pulp && /var/lib/pulp/content && /var/lib/mongodb) are not available (validation code at https://github.com/theforeman/smart-proxy-pulp/commit/7b4d074b904bedaa595ac53986f579bff14882b8#diff-774e0be2d698961baf0731e9f26cf9d7R9)
Maybe I should remove `/var/lib/pulp/content` from the read validation if pulp creates it on the fly when syncing?

ehelms commented:
Visually, on those screenshots, it says 1 GB of 1 GB used but the bar isn't full?

shlomizadok commented:
@ehelms - it is because We count in Gigabytes and that specific directory (/boot) it has 200mb / 500mb  but in GB it is 1 of 1....
I thought that GB should be default, but we can change it

ehelms commented:
Its tough, and almost circumstantial whether to use GB or not -- what about 1 or 2 decimental places? then you'd get something like 0.96 GB of 1 GB Used ? If that's not used much anywhere, we could adopt some MB logic for when its close.

shlomizadok commented:
@ehelms - agreed. I think I fetch size in kilobytes from the proxy, and manipulate according to the size

jlsherrill commented:
@shlomizadok thinking about your question around non-readable or missing directories, i think it may be best to just report that in the report from the smart proxy rather than not loading the plugin.  This gives the user the ability to see what is wrong rather the plugin just being missing.

shlomizadok commented:
@jlsherrill - we had this discussion with @witlessbird and @ohadlevy and came to the conclusion that it is best not to start the smart_proxy_pulp plugin if those directories are missing / unreadable. I would say that it should be so (maybe with the exception of pulp content dir (/var/lib/pulp/content) which may be created on the fly when pulp starts syncing. )
With the new proxy visibility a user will see that pulp is not loaded. (and with @lzap's / @witlessbird's event logging, he could see why)

Regardless, I have addressed @ehelms comment on the sizes, so now it looks like:
![image](https://cloud.githubusercontent.com/assets/433583/12492854/05788a06-c08b-11e5-9cc8-94967c8e23e3.png)

I have also added functional tests to the smart_proxies new actions. Please advise me which other tests you'd like to see on the PR and what is missing to make it mergeable :)

Thanks!

ehelms commented:
I'm slightly confused, so I'll try to recap what I think I read and you tell me where I go wrong :)

 1) install smart proxy with Pulp and pulp plugin
 2) because its a new install, /var/lib/pulp/content is missing and the smart proxy pulp plugin doesn't get loaded
 3) The smart proxy doesn't report Pulp information to the user? Or reports error logs saying Pulp plugin couldn't be loaded?
 4) User performs some action that syncs content to the smart proxy for the first time, creating /var/lib/pulp/content
 5) Somehow the smart proxy pulp plugin is reloaded on the fly and enabled?

ehelms commented:
Also - screenshot looks great, will test and re-review overall PR

shlomizadok commented:
@ehelms - I am sorry for the confusion.
Let me clear this up a little:
1. Pulp plugin on the smart proxy is validating that the following directories exist and readable:
`/var/lib/pulp`, `/var/lib/pulp/content/` & `/var/lib/mongodb`
2. If all three directories exist and readable, the Pulp plugin is loaded on the proxy and everyone is :smile: 
3. If one (or more) does not exist or not readable the Pulp plugin won't load and will log that it was not loaded due to failing to read the directory(ies)
4. It seems that `/var/lib/pulp/content` may be created on the fly when pulp starts syncing content 
(Hence my question: Shall we remove it from pulp proxy validation?)
5. Pulp proxy plugin won't reload automatically, we'd need to restart the proxy after all three directories exist.

ohadlevy commented:
On Sun, Jan 24, 2016 at 11:52 AM, Shlomi Zadok <notifications@github.com>
wrote:

> 5. Pulp proxy plugin won't reload automatically, we'd need to restart the
> proxy after all three directories exist.



IMHO if the pulp proxy feature is not running, the sync can't even start as
the proxy/capsule is not registered in the first place?

Therefore, I would probably report it, but not hard fail on it. esp on the
content directory.

shlomizadok commented:
On a separate commit: Change how to calculate space warning. Rescue if pulp service is unavailable.

shlomizadok commented:
[test] unrelated(? ) javascript error

shlomizadok commented:
@ehelms, @jlsherrill, @ohadlevy - So my followup question is:
I now understand we need to change the validation on `smart-proxy-pulp` plugin. 
Please help me figure which is the prefered approach:

* "I don't really care about `/var/lib/pulp/content` as it is created on the fly by pulp. Pulp plugin should fail loading _only_ if `/var/lib/pulp` & `/var/lib/mongodb` are missing (Hint: Pulp won't start anyhow)

Or

* "smart-proxy-pulp shouldn't fail if any of the directories is missing. If the directory is missing, skip testing its disk usage"

Thanks :small_red_triangle:

jlsherrill commented:
My vote would be for b) not failing for any directories.  To me its not the smart proxy's job to figure out if pulp is running or not, just report various statuses.  the status report should show that in a much better way if pulp is not running.

shlomizadok commented:
@jlsherrill - https://github.com/theforeman/smart-proxy-pulp/pull/4 should cover option b

jlsherrill commented on app/helpers/katello/concerns/smart_proxy_helper_extensions.rb:
```diff
@@ -0,0 +1,51 @@
+      end
+
+      def boolean_to_icon(boolean)
+        boolean = Foreman::Cast.to_bool(boolean)
```
getting an error, i think you need to prepend '::'  onto Foreman::Cast

we might should try to remove the Katello::Foreman module at some point :)

shlomizadok replied:
Weird. I didn't hit that error. Adding.

jlsherrill replied:
yeah I think its partially due to rails loading order, i've seen it pop up on previously working code.

jlsherrill commented:
ACK from me.  @ehelms any other thoughts or concerns on your side?

ehelms commented:
@shlomizadok working through testing this I noticed it handles errors kinda ugly

![screenshot from 2016-01-26 10-35-05](https://cloud.githubusercontent.com/assets/293363/12585350/af6cbb2c-c419-11e5-9388-48315a9bcac2.png)

shlomizadok commented:
@ehelms :sob: omg, it does look ugly !
Fixing :)

ehelms commented on app/views/smart_proxies/_disk_usage.html.erb:
```diff
@@ -0,0 +1,15 @@
+<div class="card-pf">
+  <div class="card-pf-body">
+    <div class="progress-description">
+      <%= disk['header'] %>
```
last comment -- adding a title attribute here would allow users to see all of the content places, sometimes I see this getting ellipsied when the content is too long:

![screenshot from 2016-01-26 10-52-27](https://cloud.githubusercontent.com/assets/293363/12585682/13f10606-c41b-11e5-9c4c-50304632f14d.png)

shlomizadok commented:
@ehelms - Should be okay now (Thanks!)

ehelms commented:
Will there be a build of the smart proxy pulp plugin going out shortly for nightly/development users to update to?

shlomizadok commented:
@ehelms,
1. New smart-proxy-pulp plugin version should be out soon (will see to that)
2. What title shall I add?

ehelms commented:
I would just add the same information, that way a user can at least hover over and see all of the disk information that might be ellipsisied

shlomizadok commented:
Added the title (good point).

ehelms commented:
Great, once tests are green and a gem of the smart proxy pulp plugin is available we'll merge this.

shlomizadok commented:
@witlessbird, will you kindly update us when a new  smart-proxy-pulp version is out?
Thanks!

ehelms commented:
Looks like this is including an already merged commit somehow, perhaps from a rebase gone awry?

shlomizadok commented:
This is after rebasing... Not sure why it was added :/

ehelms commented:
You might try cherry-picking your commit onto HEAD of master in a new branch and pushing that to your branch?

shlomizadok commented:
Messed up my git. Closing this and re-opwning a new one w/ cherry-pick. sorry
